Move continuous monitor plotting to matplotlib

maxwolffe opened the pull request:
This PR uses Matplot lib instead of gnuplot for graphing the continuous monitor. 

This script doesn't save by default and automatically opens up the plots, after which you can save by clicking save, let me know if it'd be helpful to add automatically saving, or anything else. 

Once we've decided what features we want, I can move any of our other plotting scripts to matplotlib.

kayousterhout commented:
If you could add saving by default, that would be great! Also I think @christophercanel prefers that plots aren't opened automatically (or at least that there's some option to make them not open) because he often uses the scripts to plot a bunch of experiments at a time, where it was annoying that the graphs all open immediately.  Chris, correct me if I'm misunderstanding what's useful for you!

Also, is it possible to maintain the gnuplot plotting, in addition to the new matplot lib one? It looks like this change eliminates all of the gnuplot stuff.  I think you can still (relatively easily) output the gnuplot files by outputting the values of the dictionary you created with all of the data.  I'd like to keep the gnuplot stuff if it's not too ugly to do so, since if we ever use one of these graphs in a paper, we'll likely use the gnuplot ones, so it's nice to have those around still.

Thanks for doing this!

maxwolffe commented:
@kayousterhout - I've added back in gnuplot, saving matplotlib graphs by default, and removed opening on default. I've also refactored the code a bit to conform more to PEP8 style (4 space indents, 2 newlines between function definitions, etc), let me know your thoughts.

kayousterhout commented:
I'll look at this in more detail tomorrow, but re:indentation, can you change this back to 2 spaces? Totally agree that, in retrospect, I should have used 4 spaces (we use the Google Python style guide, which similarly dictates 4 spaces) but now that all of the other python files in this repo are 2 spaces, we should keep that for consistency (unless you want to change all of those to 4 spaces -- if you do want to do that, do it in a separate PR).

maxwolffe commented:
Yep! I'll remove the PEP8 stuff for now and will add it in a separate PR.

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -3,46 +3,26 @@
 import os
 import subprocess
-import sys
+import matplotlib.pyplot as plt
```
can you use the full name (pyplot) instead of re-naming it? that generally makes it easier to read code and understand where library calls are coming from

kayousterhout replied:
Add a new line between the matplotlib imports and the other ones

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -128,93 +108,40 @@ def plot_continuous_monitor(filename, open_graphs=False):
+    }
+    continuous_monitor_data.append(data)
+    if use_gnuplot:
+      write_data(out_file, gnuplot_data_line(data))
```
Can you move this functionality to the plot_gnuplot function, so that plot_gnuplot makes out_file, writes the data to it, and then does the plotting? That way all of the code related to the gnuplot plotting is in one place.

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -226,9 +153,171 @@ def get_util_for_disk(disk_utils, disk):
   return None
 
+def gnuplot_data_line(data_hash):
+    """ Helper function to order continuous montior data for gnuplot to use """
```
monitor

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -226,9 +153,171 @@ def get_util_for_disk(disk_utils, disk):
 
+def gnuplot_data_line(data_hash):
+    """ Helper function to order continuous montior data for gnuplot to use """
+    return [data_hash['time'],
```
This looks very possibly error prone in the future.  Instead, what about changing line 111 to construct a list instead of a dictionary (so just change the curly braces to square ones). Then, you can use dict() to make it into a dictionary to pass into the matplot lib stuff, and you can do something like [a[0] for a in list] to get the data for gnuplot.

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -226,9 +153,171 @@ def get_util_for_disk(disk_utils, disk):
+
+def plot_gnuplot(file_prefix, open_graphs):
+  """
+  Creates modified copies of gnuplot files, and associated pdf graphs for several
```
What do you mean modified copies? What about "Creates gnuplot files that can be used to generate plots with data for various continuous monitor attributes, and uses those files to generate PDFs of each plot."

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -226,9 +153,171 @@ def get_util_for_disk(disk_utils, disk):
+  for attribute in attributes:
+    plot_gnuplot_attribute(file_prefix, attribute, scripts_dir)
+
+  if open_graphs:
```
move this code to plot_gnuplot_attribute? (then you already have the pdf name)

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -226,9 +153,171 @@ def get_util_for_disk(disk_utils, disk):
+    """
+    Creates a matplotlib graph using continuous monitor data.
+    Time is the x axis and data corresponding to each parameter is used to
+    generate a new line on the line graph.
```
How about "Creates a matplotlib graph that plots a line for each parameter in params_to_plot as a function of time."

kayousterhout commented:
This plot file is getting a little unwieldy... can you separate the matplotlib stuff into its own file that gets called from plot_continuous_monitor.py? Maybe matplotlib_helper.py?

maxwolffe commented:
Thanks for the suggestions @kayousterhout. I've broken the matplotlib and gnuplot plotting functions into their own files and made the changes that you suggested. Let me know if you have any more comments.

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -1,48 +1,13 @@
-import subprocess
-import sys
+import plot_gnuplot
+import plot_matplotlib
```
alphabetization

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -129,106 +94,66 @@ def plot_continuous_monitor(filename, open_graphs=False):
+      ('time', time - start),
+      ('xvdf utilization', xvdf_total_utilization),
+      ('xvdb utilization', xvdb_total_utilization),
+      ('cpu utilization', cpu_total / 8.0),  # Cores in machine?
```
8 is the cores per machine, as you guessed (might be nice to move this to a constant)

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -129,106 +94,66 @@ def plot_continuous_monitor(filename, open_graphs=False):
+      ('xvdb utilization', xvdb_total_utilization),
+      ('cpu utilization', cpu_total / 8.0),  # Cores in machine?
+      ('bytes received', bytes_received / 125000000.),
+      ('bytes transmitted', bytes_transmitted / 125000000.),  # what is the 125000000 magic number?
```
125000000 is the number of bytes in a gigabit (these scripts assume the machines have 1Gbps network links)

maxwolffe replied:
Great! I'll move both of these to constants.

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -129,106 +94,66 @@ def plot_continuous_monitor(filename, open_graphs=False):
+    "-o", "--open_graphs", help="open generated graphs", action="store_true")
+  parser.add_argument(
+    "-g", "--gnuplot", help="generate graphs with gnuplot", action="store_true")
+  parser.set_defaults(gnuplot=False, open_graphs=False)
```
Can you leave the old default mechanism (makes the args easier to read, since the default is close to the definition)

maxwolffe replied:
Yep!

kayousterhout commented on plot_matplotlib.py:
```diff
@@ -0,0 +1,76 @@
+    time = continuous_monitor_col('time', cm_data)
+    for key in params_to_plot:
+      handle, = pyplot.plot(time, continuous_monitor_col(key, cm_data),
+                         label=key)
```
indentation

kayousterhout commented on plot_matplotlib.py:
```diff
@@ -0,0 +1,76 @@
+    handles = []
+    time = continuous_monitor_col('time', cm_data)
+    for key in params_to_plot:
+      handle, = pyplot.plot(time, continuous_monitor_col(key, cm_data),
```
What does the "," mean here?

maxwolffe replied:
It's the same as pyplot.plot()[0]. This was more a case of copy-pasta. I'll change it to [0]

kayousterhout commented:
This looks good! Just a few last small things

maxwolffe commented:
Thanks for the comments Kay. When you think this is good to go, I can squash the commits so it's a little cleaner.

kayousterhout commented on plot_continuous_monitor.py:
```diff
@@ -4,7 +4,9 @@
 BYTES_PER_GIGABYTE = float(1024 * 1024 * 1024)
-
+BYTES_PER_KILOBYTE = 1024 * 1024
+BYTES_PER_GIGABIT = 125000000.
```
tiny nit: can you do BYTES_PER_GIGABYTE / 8?

maxwolffe replied:
oh yeah, absolutely.

kayousterhout commented on plot_matplotlib.py:
```diff
@@ -50,8 +50,8 @@ def plot_params(params_to_plot, title):
-      handle, = pyplot.plot(time, continuous_monitor_col(key, cm_data),
-                         label=key)
+      handle = pyplot.plot(time, continuous_monitor_col(key, cm_data),
+                            label=key)[0]
```
This indentation still looks like it's off by 1 space? am i crazy?

maxwolffe replied:
Nope! Not crazy. I fixed it, then tested at [0] addition to make sure it behaved the same, then forgot to change it back. doh.

kayousterhout commented:
Merged!
